File: ts/src/doublures/create-user.service.ts
export class User {
  constructor(
    readonly id: string,
    readonly name: string,
    readonly email: any
  ) {}
}

export interface IUserRepository {
  insert(name: string, email: string): string;

  get(id: string): User | null;
}

export class UserValidationService {
  sendActivationLinkToUser(id: string): void {
    // Do something potentially complex
    throw Error('AH AH YOU FAIL');
  }
}

export class NotificationService {
  sendNotification(to: string, message: string): void {
    // Do something we don't care for now
  }
}

export class CreateUserService {
  constructor(
    private userRepository: IUserRepository,
    private userValidationService: UserValidationService,
    private notificationService: NotificationService
  ) {}

  createUser(name: string, email: string): string {
    const id = this.userRepository.insert(name, email);
    this.userValidationService.sendActivationLinkToUser(id);
    return id;
  }
}

File: java/src/main/java/test/doubles/CreateUserService.java
package test.doubles;

public class CreateUserService {

    private IUserRepository userRepository;
    private UserValidationService userValidationService;
    private NotificationService notificationService;

    CreateUserService(IUserRepository userRepository, UserValidationService userValidationService, NotificationService notificationService ) {
        this.userRepository = userRepository;
        this.userValidationService = userValidationService;
        this.notificationService = notificationService;
    }

    String createUser(String name, String email) {
        var id = this.userRepository.insert(name, email);
        this.userValidationService.sendActivationLinkToUser(id);
        return id;
    }
}
